### PR TITLE
Some improvements for KV caching

### PR DESCRIPTION
**Pull request by mseeger**
- Ensure that KVCache buffers are only as large as config.n_query_groups
- Shrink buffers returned by KVCache to just cover input_pos entries
- Clean up children of classes in `model.py`, in particular remove `forward` copies

### CODE REVIEW

**Comment by mseeger** [comment]
Can somebody help with failing tests? I don't understand why tests for Windows are failing, but pass for all other systems. And I also don't understand why the GPU tests are failing.

**Comment by Andrei-Aksionov** [comment]
Hello @mseeger 

Thank you for another PR.

> Can somebody help with failing tests? I don't understand why tests for Windows are failing

yeah, there is always something with Windows.
Perhaps we are lucky and a simple torch bump from #1893 might help 🤷 

> And I also don't understand why the GPU tests are failing.

I'll check it tomorrow.

File: litgpt/adapter.py [hunk 1]
```diff
@@ -91,30 +62,10 @@ def _init_weights(self, module: nn.Module) -> None:
...
-                " non-parallel residual and shared attention norm."
-            )
-        self.norm_1 = config.norm_class(config.n_embd, eps=config.norm_eps)
+        super().__init__(config, block_idx, skip_attn_mlp=True)
         self.attn = CausalSelfAttention(config, block_idx)
-        self.post_attention_norm = (
-            config.norm_class(config.n_embd, eps=config.norm_eps) if config.post_attention_norm else nn.Identity()
-        )
-        self.norm_2 = None if config.shared_attention_norm else config.norm_class(config.n_embd, eps=config.norm_eps)
         self.mlp = config.mlp_class(config)
```
[the remark this answers is not in the document]
Comment (Andrei-Aksionov) [reply litgpt/adapter.py:68]: Oh, now I see.
You provide `skip_attn_mlp=True`, so in the base class self.attn and self.mlp are assigned to None.
I'm ok with reassigning it here.
Otherwise it might look a bit confusing, why a child class dictates the implementation of the base class.

I think it's better to drop `skip_attn_mlp` at all.
Yes, it will lead to a bit longer initialization, but the code in the base class will be a bit cleaner.

File: litgpt/model.py [hunk 2]
```diff
@@ -304,25 +365,28 @@ def forward(
         # - C*         | attentions's embeddings size
         # - nh_(q,k,v) | number of heads for query, key and value
         # - hs         | head size
-
-        B, T, C = x.size()
+        n_head = self.config.n_head
+        n_query_groups = self.config.n_query_groups
+        rope_n_elem = self.config.rope_n_elem
+        head_size = self.config.head_size
+        B, T, C = x.size()  # batch size, sequence length, embedding dimensionality (n_embd)
```
Comment (Andrei-Aksionov) [inline litgpt/model.py:372]:
```suggestion
        head_size = self.config.head_size
        n_head = self.config.n_head
        n_query_groups = self.config.n_query_groups
        rope_n_elem = self.config.rope_n_elem
        B, T, C = x.size()  # batch size, sequence length, embedding dimensionality (n_embd)
```

just a small reordering.

File: litgpt/model.py [hunk 3]
```diff
@@ -332,22 +396,28 @@ def forward(
...
+        k_roped = apply_rope(k[..., : rope_n_elem], cos, sin)
+        q = torch.cat((q_roped, q[..., rope_n_elem :]), dim=-1)  # (B, nh_q, T, hs)
+        k = torch.cat((k_roped, k[..., rope_n_elem :]), dim=-1)  # (B, nh_k, T, hs)
 
         # Apply kv-cache during inference.
         if input_pos is not None:
             if not isinstance(self.kv_cache, KVCache):
                 raise TypeError("You need to call `gpt.set_kv_cache()`")
             k, v = self.kv_cache(input_pos, k, v)
+            if input_pos_maxp1 is not None:
```
Comment (Andrei-Aksionov) [inline litgpt/model.py:409]: Shouldn't it be in `KVCache.forward`?

File: litgpt/model.py [hunk 4]
```diff
@@ -375,7 +446,7 @@ def forward(
         y = self.scaled_dot_product_attention(q, k, v, mask)
 
         # Re-assemble all head outputs side by side.
-        y = y.reshape(B, T, self.config.head_size * self.config.n_head)
+        y = y.flatten(start_dim=-2, end_dim=-1)
```
Comment (Andrei-Aksionov) [inline litgpt/model.py:449]: Here it's a bit over-engineered 🙂.

Overall, I think that the existing one is easier to read and gives you info of the final shape.

File: litgpt/model.py [hunk 5]
```diff
@@ -642,6 +744,11 @@ def do_softcapping(x: torch.Tensor, thresh: float) -> torch.Tensor:
 
 
 class KVCache(nn.Module):
+    """
+    Buffers `k`, `v` have shape
+    `(batch_size, n_query_groups, max_seq_length, head_size)`. Call dimension
+    2 the sequence dimension.
```
Comment (Andrei-Aksionov) [inline litgpt/model.py:750]:
```
Call dimension
    2 the sequence dimension.
```
Is this needed here?

File: litgpt/model.py [hunk 6]
```diff
@@ -654,13 +761,27 @@ def __init__(
         self.register_buffer("v", torch.zeros(v_shape, device=device, dtype=dtype), persistent=False)
 
     def forward(self, input_pos: torch.Tensor, k: torch.Tensor, v: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        """
+        Writes new values `k`, `v` into cache at positions in `input_pos` along
+        the sequence dimension. Their batch size `bs` must be `<= batch_size`.
+        Returns the complete buffers, selected to batch size `bs`.
+
```
Comment (Andrei-Aksionov) [inline litgpt/model.py:768]: Writes new values `k` and `v` into the cache at the positions specified by `input_pos` along the sequence dimension (`max_seq_length`). The batch size of `k` and `v` (`bs`) must be smaller or equal to KVCache batch size. Returns the full buffers, adjusted to the batch size `bs`.

**Comment by Andrei-Aksionov** [comment]
Hello @mseeger 

It's quite a PR 🫠 🙂.
I left a couple of comments.
Overall it looks really good. I like how PEFT variants now look like, allows focusing on differences easily now 😊.

(I'll take a look why GPU tests are failing later.)

File: litgpt/adapter.py [hunk 7]
```diff
@@ -28,56 +28,27 @@ class Config(BaseConfig):
 
 
 class GPT(BaseModel):
-    """The implementation is identical to `litgpt.model.GPT` with the exception that
-    the `Block` saves the layer index and passes it down to the attention layer."""
-
+    # Copy&paste from :class:`model.GPT`
     def __init__(self, config: Config) -> None:
```
[the remark this answers is not in the document]
Comment (mseeger) [reply litgpt/adapter.py:32]: I think the problem is that `transformer.h` is creating `Block` objects of the kind defined here. The superclass constructor would use the wrong `Block` type, no?

File: litgpt/adapter.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mseeger) [reply litgpt/adapter.py:68]: OK, makes sense

File: litgpt/adapter.py [hunk 8]
```diff
@@ -28,56 +28,27 @@ class Config(BaseConfig):
...
-            # chunk the lm head logits to reduce the peak memory used by autograd
-            return [self.lm_head(x_i) for x_i in x.split(lm_head_chunk_size, dim=1)]
-        x = self.lm_head(x)  # (b, t, vocab_size)
-        if self.config.final_logit_softcapping is not None:
-            x = torch.tanh(x / self.config.final_logit_softcapping) * self.config.final_logit_softcapping
-        return x
+        self.max_seq_length = self.config.block_size
 
     @classmethod
     def from_name(cls, name: str, **kwargs: Any) -> Self:
```
[the remark this answers is not in the document]
Comment (mseeger) [reply litgpt/adapter.py:54]: Should I do that? I don't have the full overview here.

File: litgpt/adapter_v2.py [hunk 9]
```diff
@@ -64,54 +65,27 @@ def reset_parameters(self) -> None:
...
 class GPT(BaseModel):
+    # Copy&paste from :class:`model.GPT`
     def __init__(self, config: Config) -> None:
-        # Skip the parent class __init__ altogether and replace it to avoid useless allocations
         nn.Module.__init__(self)
         assert config.padded_vocab_size is not None
         self.config = config
 
-        self.lm_head = AdapterV2Linear(config.n_embd, config.padded_vocab_size, bias=config.lm_head_bias)
+        self.lm_head = AdapterV2Linear(
```
[the remark this answers is not in the document]
Comment (mseeger) [reply litgpt/adapter_v2.py:74]: Again, note that the `Block` type is different here.

[the remark this answers is not in the document]
Comment (mseeger) [reply litgpt/adapter_v2.py:74]: The only way to be able to call the superclass `__init__` would be to introduce these factory methods

File: litgpt/model.py [hunk 10]
```diff
@@ -74,21 +83,41 @@ def _init_weights(self, module: nn.Module) -> None:
...
+        If `input_pos` is given, the KV cache is used for K, V vectors at
+        positions smaller than entries in `input_pos`. Consider passing the
+        maximum of its entries plus 1 in `input_pos_maxp1` if this value is
+        determined by your forward algorithm anyway. If so, the multi-head
+        attention computation is sped up by slicing the KV cache buffers
+        accordingly. If `input_pos_maxp1` is not given, the MHA computation
+        is done with complete KV cache buffers (length 'max_seq_length`), and
+        masking is applied. We cannot compute `input_pos_maxp1` from
+        `input_pos`, since this would induce graph breaks and prevent graph
+        compilation.
```
[the remark this answers is not in the document]
Comment (mseeger) [reply litgpt/model.py:103]: OK

File: litgpt/model.py (hunk 3 above)
Comment (mseeger) [reply litgpt/model.py:409]: I wanted the `KVCache` not to have to bother about this feature.

File: litgpt/model.py (hunk 4 above)
Comment (mseeger) [reply litgpt/model.py:449]: OK

File: litgpt/model.py (hunk 5 above)
Comment (mseeger) [reply litgpt/model.py:750]: Removed

**Comment by mseeger** [comment]
OK, I reacted to comments. I also did a small change in `lora.py`, where the `CausalSelfAttention.__init__` was still copy and paste, now it calls the superclass init.

File: litgpt/adapter.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Andrei-Aksionov) [reply litgpt/adapter.py:32]: Yeah, you are totally right.
My bad 🙂

File: litgpt/adapter.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Andrei-Aksionov) [reply litgpt/adapter.py:54]: No, again, the same issue. 
If to remove it and use the one from the base class:
```python
@classmethod
    def from_name(cls, name: str, **kwargs: Any) -> Self:
        return cls(Config.from_name(name, **kwargs))
```
it will create an instance of the base class with the wrong variant of Block, ...

File: litgpt/adapter_v2.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Andrei-Aksionov) [reply litgpt/adapter_v2.py:72]: Agree.
Again, my bad 🙂

File: litgpt/model.py [hunk 11]
```diff
@@ -213,15 +264,15 @@ def __init__(self, config: Config, block_idx: int) -> None:
...
         self.post_attention_norm = (
             config.norm_class(config.n_embd, eps=config.norm_eps) if config.post_attention_norm else nn.Identity()
         )
         self.norm_2 = None if config.shared_attention_norm else config.norm_class(config.n_embd, eps=config.norm_eps)
-        self.mlp = config.mlp_class(config)
         self.post_mlp_norm = (
             config.norm_class(config.n_embd, eps=config.norm_eps) if config.post_mlp_norm else nn.Identity()
         )
+        self.attn = CausalSelfAttention(config, block_idx)
+        self.mlp = config.mlp_class(config)
```
Comment (Andrei-Aksionov) [inline litgpt/model.py:275]: Could you move them back where they were before?

File: litgpt/model.py (hunk 3 above)
Comment (Andrei-Aksionov) [reply litgpt/model.py:404]: Ok. Let's leave it here.

**Comment by Andrei-Aksionov** [comment]
> OK, I reacted to comments. I also did a small change in lora.py, where the CausalSelfAttention.__init__ was still copy and paste, now it calls the superclass init.

Cool, we are almost there 🙂.
There are a couple of unresolved comments left.

On my side I'll try to find and fix issues with failing GPU tests, hopefully this year 😃.

File: litgpt/model.py (hunk 11 above)
Comment (mseeger) [reply litgpt/model.py:275]: OK

File: litgpt/model.py [hunk 12]
```diff
@@ -101,31 +130,49 @@ def forward(self, idx: torch.Tensor, input_pos: Optional[torch.Tensor] = None) -
...
             # unsqueeze to have a batch dimension
             cos = self.cos[:T].unsqueeze(0)
             sin = self.sin[:T].unsqueeze(0)
             # `cos`, `sin` have shape (1, T, config.rope_n_elem)
             mask = None  # defaults to causal mask
+            input_pos_maxp1 = None
 
         x = self.transformer.wte(idx)  # token embeddings of shape (B, T, n_embd)
         if self.config.scale_embeddings:
-            x = x * torch.tensor(self.config.n_embd**0.5, dtype=x.dtype)
```
[the remark this answers is not in the document]
Comment (mseeger) [reply litgpt/model.py:120]: OK

File: litgpt/model.py [hunk 13]
```diff
@@ -101,31 +130,49 @@ def forward(self, idx: torch.Tensor, input_pos: Optional[torch.Tensor] = None) -
...
 
         for block in self.transformer.h:
-            x = block(x, cos, sin, mask, input_pos)
+            x = block(x, cos, sin, mask, input_pos, input_pos_maxp1)
         x = self.transformer.ln_f(x)
-        x = self.lm_head(x)  # (B, T, padded_vocab_size)
-        if self.config.final_logit_softcapping is not None:
-            x = do_softcapping(x, self.config.final_logit_softcapping)
-        return x
+        extra_map = (
```
[the remark this answers is not in the document]
Comment (mseeger) [reply litgpt/model.py:163]: OK

File: litgpt/model.py [hunk 14]
```diff
@@ -101,31 +130,49 @@ def forward(self, idx: torch.Tensor, input_pos: Optional[torch.Tensor] = None) -
...
+            partial(do_softcapping, thresh=self.config.final_logit_softcapping)
+            if self.config.final_logit_softcapping is not None
+            else nn.Identity()
+        )
+        if lm_head_chunk_size > 0:
```
[the remark this answers is not in the document]
Comment (mseeger) [reply litgpt/model.py:168]: It would be nice if this could be removed in the future, indeed.

**Comment by Andrei-Aksionov** [comment]
Overall, the issue with GPU+Thunder is something specific to the latter.
I'll merge the PR as is and later discuss it with Thunder team.

Thanks again for the PR (and for the patience 😊).

Happy New Year! 🚀

**Comment by t-vi** [comment]
@mseeger unfortunately the "Shrink buffers returned by KVCache to just cover input_pos entries" does not work as intended: Introducing data-dependent control flow (i.e. here making a tensor size out of a tensor value) typically breaks compilation just as much as using CPU integers.

The other aspect here is that in my experience, the performance impact of this has been extremely limited if the attention implementation works reasonably, so I would have a tendency to revert this part of the change.

**Comment by mseeger** [comment]
Hello, my original PR did not make a tensor size out of a tensor value. In fact, this is why `input_pos_maxp1` must be an extra argument (of int type) instead of just computing the max of `input_pos`.

But sure, if you like to revert this (the other changes I think are useful). I am already working on something that would make this obsolete.

**Comment by mseeger** [comment]
My intent is that `LitGPT` supports modern KV caching and sparse attention. That would be something that for example Hugging Face does not do properly. They support some KV caching, but only for a few models.

**Comment by mseeger** [comment]
With what I am preparing, KV caches would receive a proper abstract definition (with a very simple implementation for what is currently given, namely the exact KV cache), and this `input_pos_maxp1` would be removed.

**Comment by t-vi** [comment]
```python
 if input_pos_maxp1 > self.max_seq_length:
    raise ValueError(f"Positions in 'input_pos' must be in [0,{self.max_seq_length})")
mask = mask[..., :input_pos_maxp1]
```

So in these lines, if `input_pos_maxp1` is a tensor, the if is data-dependent control-flow because it depends on the value of the tensor. I'm not sure if the line is per se useful.
The second line, using the tensor for slicing, makes mask of size `min(old_size, input_pos_maxp1)` which depends on the value of the tensor.

I know these things are subtle. In #1912 we specifically drop the input_pos_maxp1 this from the generate when dealing with ThunderModules.
What we found useful (and I would prefer over the current implementation) is instead of slicing the tensors, we pass the input_pos_maxp1 and the full tensors to the attention function and then that decide not to look at later parts of the tensors. This may also be advantageous in case of "funny sizes" (not divisible by whatever power of two the attention kernel wants for loading things into memory) because it seems not so great style to access tensors out of bounds even if you know they are slices of a larger ones. We could have a default attention function that does the slicing in eager mode and then compilers can override it as needed.

**Comment by t-vi** [comment]
(And needless to add, I do greatly appreciate your work on KVCaches in LitGPT.)
